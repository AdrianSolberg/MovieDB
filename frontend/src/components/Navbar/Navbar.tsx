import NavbarOverlay from "./NavbarOverlay";
import UserDropdown from "./UserDropdown";

export type NavbarProps = { name: string; href: string }[];

const Navbar = () => {
    const links: NavbarProps = [
        { name: "Movies", href: "/movies" },
        { name: "Feed", href: "/feed" },
    ];

    return (
        <header className="sticky top-0 left-0 z-50 bg-muted shadow-sm w-dvw flex items-center p-4">
            <nav className="flex-1 flex justify-center items-center gap-10">
                <a href="/" className="text-2xl font-bold text-primary">
                    Logo
                </a>
                <ul className="hidden sm:flex gap-4 text-lg">
                    {links.map((link) => (
                        <li key={link.name}>
                            <a
                                href={link.href}
                                className="hover:text-primary text-foreground transistion-colors"
                            >
                                {link.name}
                            </a>
                        </li>
                    ))}
                </ul>
            </nav>
            <div className="absolute right-4 flex items-center gap-2">
                <UserDropdown />
                <NavbarOverlay links={links} />
            </div>
        </header>
    );
};

export default Navbar;
