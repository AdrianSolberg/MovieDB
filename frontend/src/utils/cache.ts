import { makeVar } from "@apollo/client";
import { Filters, SortingType } from "@/types/movieTypes";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);


const storedFilters = sessionStorage.getItem("filters");
const initialFilters: Filters = storedFilters ? JSON.parse(storedFilters) : {
    Genre: [],
    Rating: [],
    Decade: [],
    Status: [],
    Runtime: [],
};
export const filtersVar = makeVar<Filters>(initialFilters);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);

export const usernameVar = makeVar<string>(
    localStorage.getItem("username") ?? "Guest"
);
