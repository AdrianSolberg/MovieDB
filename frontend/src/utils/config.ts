const BACKEND_PORT = 3001;

const BACKEND_URL =
    import.meta.env.MODE === "production"
        ? `http://it2810-26.idi.ntnu.no:${BACKEND_PORT}`
        : `http://localhost:${BACKEND_PORT}`;

export { BACKEND_URL };
