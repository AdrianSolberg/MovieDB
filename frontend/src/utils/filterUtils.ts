import { Filters, FiltersInput } from "@/types/__generated__/types";

export function getFiltersAsInput(filters: Filters): FiltersInput {
    return {
        Decade: filters.Decade.map((filter) => filter.name),
        Genre: filters.Genre.map((filter) => filter.name),
        Rating: filters.Rating.map((filter) => filter.name),
        Status: filters.Status.map((filter) => filter.name),
        Runtime: filters.Runtime.map((filter) => filter.name),
    };
}
