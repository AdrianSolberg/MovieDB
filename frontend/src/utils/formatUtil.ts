export const formatDate = (date: Date) => {
    return date.toLocaleDateString("en-US", {
        year: "numeric",
        month: "long",
        day: "numeric",
        hour: "2-digit",
        minute: "2-digit",
    });
};

export const formatNumber = (n: number) => {
    return n.toLocaleString("no-NO");
};

export const formatCurrency = (value: number) => {
    return value.toLocaleString("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
    });
};
