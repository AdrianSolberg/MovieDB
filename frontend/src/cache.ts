import { makeVar } from "@apollo/client";
import { SortingType } from "./utils/searchSortAndFilter";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);

export const filtersVar = makeVar<{ [key: string]: string[] }>(
    JSON.parse(sessionStorage.getItem("filters") ?? "{}")
);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);
