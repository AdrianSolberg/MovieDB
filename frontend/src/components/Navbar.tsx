import { Link } from "react-router-dom";
import logo from "../assets/logo.png";
import NavbarOverlay from "./NavbarOverlay";
import UserDropdown from "./UserDropdown";
import { useReactiveVar } from "@apollo/client";
import { usernameVar } from "@/utils/cache";

export type NavbarProps = { name: string; href: string }[];

const Navbar = () => {
    const username = useReactiveVar(usernameVar);
    const links: NavbarProps = [
        { name: "Activity", href: "/activity" },
        ...(username != "Guest"
            ? [{ name: "My Reviews", href: "/myReviews" }]
            : []),
    ];

    return (
        <header className="flex w-full items-center justify-between bg-gradient-to-r from-muted-foreground to-primary p-4">
            <nav className="flex items-center gap-10 text-background">
                <Link to="/" className="flex flex-row gap-3 text-2xl font-bold">
                    <img src={logo} alt="logo" className="h-8" />
                    MovieDB
                </Link>
                <ul className="hidden gap-4 text-lg sm:flex">
                    {links.map((link) => (
                        <li key={link.name}>
                            <Link to={link.href} className="hover:underline">
                                {link.name}
                            </Link>
                        </li>
                    ))}
                </ul>
            </nav>
            <div className="flex items-center">
                <UserDropdown />
                {links.length > 0 && (
                    <div className="ml-2">
                        <NavbarOverlay links={links} />
                    </div>
                )}
            </div>
        </header>
    );
};

export default Navbar;
