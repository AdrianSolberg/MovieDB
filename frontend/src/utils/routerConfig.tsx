import Navbar from "@/components/Navbar";
import ActivityPage from "@/pages/ActivityPage";
import HomePage from "@/pages/HomePage";
import MovieDetailPage from "@/pages/MovieDetailPage";
import MyReviewsPage from "@/pages/MyReviewsPage";
import NotFoundPage from "@/pages/NotFoundPage";
import { Toaster } from "@/shadcn/components/ui/sonner";
import { Outlet, RouteObject } from "react-router-dom";

const routerConfig: RouteObject[] = [
    {
        path: "/",
        element: (
            <>
                <Navbar />
                <div id="main-content" tabIndex={-1}>
                    <Outlet />
                </div>
                <Toaster position="bottom-right" />
            </>
        ),
        errorElement: <NotFoundPage />,
        children: [
            {
                path: "",
                element: <HomePage />,
            },
            {
                path: "movie/:movieId",
                element: <MovieDetailPage />,
            },
            {
                path: "activity",
                element: <ActivityPage />,
            },
            {
                path: "myReviews",
                element: <MyReviewsPage />,
            },
        ],
    },
];

export default routerConfig;
