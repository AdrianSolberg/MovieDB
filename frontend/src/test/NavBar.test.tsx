import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";
import Navbar from "@/components/Navbar";
import { vi } from "vitest";
import { usernameVar } from "@/utils/cache";

vi.mock("../components/UserDropdown", () => ({
    default: vi.fn(() => <div data-testid="user-dropdown" />),
}));

vi.mock("../components/NavbarOverlay", () => ({
    default: vi.fn(() => <div data-testid="navbar-overlay" />),
}));

vi.mock("@/utils/cache", () => ({
    usernameVar: vi.fn(),
}));

vi.mock("@apollo/client", async () => {
    const original = await vi.importActual("@apollo/client");
    return {
        ...original,
        useReactiveVar: vi.fn().mockImplementation((varFn) => varFn()),
    };
});

describe("Navbar", () => {
    it("matches snapshot for default Navbar", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <Navbar />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("successfully renders Navbar", async () => {
        vi.mocked(usernameVar).mockReturnValue("Guest");
        render(
            <MemoryRouter>
                <Navbar />
            </MemoryRouter>
        );

        expect(screen.getByAltText("logo")).toBeInTheDocument();
        expect(screen.getByText("MovieDB")).toBeInTheDocument();
        expect(screen.getByText("MovieDB")).toHaveAttribute("href", "/");

        expect(screen.getByText("Activity")).toBeInTheDocument();
        expect(screen.getByText("Activity")).toHaveAttribute(
            "href",
            "/activity"
        );

        expect(screen.queryByText("My Reviews")).toBeNull();

        expect(screen.getByTestId("user-dropdown")).toBeInTheDocument();
        expect(screen.getByTestId("navbar-overlay")).toBeInTheDocument();
    });

    it("shows 'My Reviews' for logged in users", async () => {
        vi.mocked(usernameVar).mockReturnValue("test_user");
        render(
            <MemoryRouter>
                <Navbar />
            </MemoryRouter>
        );

        expect(screen.getByText("My Reviews")).toBeInTheDocument();
        expect(screen.getByText("My Reviews")).toHaveAttribute(
            "href",
            "/myReviews"
        );
    });
});
