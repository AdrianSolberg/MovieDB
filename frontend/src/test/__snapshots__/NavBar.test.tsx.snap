// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > matches snapshot for default Navbar 1`] = `
<DocumentFragment>
  <header
    class="flex w-full items-center justify-between bg-gradient-to-r from-muted-foreground to-primary p-4"
  >
    <nav
      class="flex items-center gap-10 text-background"
    >
      <a
        class="sr-only left-0 top-0 bg-primary font-bold focus:not-sr-only focus:absolute focus:m-2 focus:p-2"
        href="#main-content"
      >
        Skip to main content
      </a>
      <a
        class="flex flex-row gap-3 text-2xl font-bold"
        href="/"
      >
        <img
          alt="logo"
          class="h-8"
          src="/project2/src/assets/logo.png"
        />
        MovieDB
      </a>
      <ul
        class="hidden gap-4 text-lg sm:flex"
      >
        <li>
          <a
            class="hover:underline"
            href="/activity"
          >
            Activity
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="flex items-center"
    >
      <button
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="user"
        class="justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center space-x-2"
        data-state="closed"
        id="radix-:r0:"
        type="button"
      >
        <svg
          class="lucide lucide-user h-4 w-4"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
          />
          <circle
            cx="12"
            cy="7"
            r="4"
          />
        </svg>
        <span
          class="hidden truncate sm:flex sm:max-w-32"
        >
          Guest
        </span>
      </button>
      <div
        class="ml-2"
      >
        <button
          aria-controls="radix-:r2:"
          aria-expanded="false"
          aria-haspopup="dialog"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10 sm:hidden"
          data-state="closed"
          type="button"
        >
          <svg
            class="lucide lucide-menu"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="4"
              x2="20"
              y1="12"
              y2="12"
            />
            <line
              x1="4"
              x2="20"
              y1="6"
              y2="6"
            />
            <line
              x1="4"
              x2="20"
              y1="18"
              y2="18"
            />
          </svg>
          <span
            class="sr-only"
          >
            Toggle navigation menu
          </span>
        </button>
        <h2
          class="text-lg font-semibold text-foreground"
          id="radix-:r3:"
        />
        <p
          class="text-sm text-muted-foreground"
          id="radix-:r4:"
        />
      </div>
    </div>
  </header>
</DocumentFragment>
`;
